vms/platformvm/txs/executor/standard_tx_executor.go

```diff
@@ -959,6 +957,142 @@ func (e *standardTxExecutor) RegisterSubnetValidatorTx(tx *txs.RegisterSubnetVal
+
+	// Parse the warp message.
+	warpMessage, err := warp.ParseMessage(tx.Message)
+	if err != nil {
+		return err
+	}
+	addressedCall, err := payload.ParseAddressedCall(warpMessage.Payload)
+	if err != nil {
+		return err
+	}
+	msg, err := message.ParseSubnetValidatorWeight(addressedCall.Payload)
+	if err != nil {
+		return err
+	}
+	if err := msg.Verify(); err != nil {
+		return err
+	}
+
+	// Verify that the message contains a valid nonce for a current validator.
+	sov, err := e.state.GetSubnetOnlyValidator(msg.ValidationID)
+	if err != nil {
+		return fmt.Errorf("%w: %w", errCouldNotLoadSoV, err)
+	}
+	if msg.Nonce < sov.MinNonce {
+		return fmt.Errorf("%w %d must be at least %d", errWarpMessageContainsStaleNonce, msg.Nonce, sov.MinNonce)
+	}
+
+	// Verify that the warp message was sent from the expected chain and
+	// address.
+	if err := verifyL1Conversion(e.state, sov.SubnetID, warpMessage.SourceChainID, addressedCall.SourceAddress); err != nil {
+		return err
+	}
+
+	txID := e.tx.ID()
+
+	// Check if we are removing the validator.
+	if msg.Weight == 0 {
+		// Verify that we are not removing the last validator.
+		weight, err := e.state.WeightOfSubnetOnlyValidators(sov.SubnetID)
+		if err != nil {
+			return fmt.Errorf("could not load SoV weights: %w", err)
+		}
+		if weight == sov.Weight {
+			return errRemovingLastValidator
+		}
+
+		// If the validator is currently active, we need to refund the remaining
+		// balance.
+		if sov.EndAccumulatedFee != 0 {
+			var remainingBalanceOwner message.PChainOwner
+			if _, err := txs.Codec.Unmarshal(sov.RemainingBalanceOwner, &remainingBalanceOwner); err != nil {
+				return fmt.Errorf("%w: remaining balance owner is malformed", errStateCorruption)
+			}
+
+			accruedFees := e.state.GetAccruedFees()
+			if sov.EndAccumulatedFee <= accruedFees {
+				// This check should be unreachable. However, it prevents AVAX
+				// from being minted due to state corruption. This also prevents
+				// invalid UTXOs from being created (with 0 value).
+				return fmt.Errorf("%w: validator should have already been disabled", errStateCorruption)
+			}
+			remainingBalance := sov.EndAccumulatedFee - accruedFees
+
+			utxo := &avax.UTXO{
+				UTXOID: avax.UTXOID{
+					TxID:        txID,
+					OutputIndex: uint32(len(tx.Outs)),
+				},
+				Asset: avax.Asset{
+					ID: e.backend.Ctx.AVAXAssetID,
+				},
+				Out: &secp256k1fx.TransferOutput{
+					Amt: remainingBalance,
+					OutputOwners: secp256k1fx.OutputOwners{
+						Threshold: remainingBalanceOwner.Threshold,
+						Addrs:     remainingBalanceOwner.Addresses,
+					},
+				},
+			}
+			e.state.AddUTXO(utxo)
+		}
+	}
+
+	// If the weight is being set to 0, it is possible for the nonce increment
+	// to overflow. However, the validator is being removed and the nonce
+	// doesn't matter. If weight is not 0, [msg.Nonce] is enforced by
+	// [msg.Verify()] to be less than MaxUInt64 and can therefore be incremented
+	// without overflow.
+	sov.MinNonce = msg.Nonce + 1
```

Comment:
does that mean we can increase the MinNonce with the nonce in the msg? I.e it does not need to be a monotonic increase? I guess a higher nonce could invalidate all others?

Reply:
Yes. We explicitly allow this so that consecutive messages can be combined into 1.

For example, say we
1. `SetWeight(nonce=5, weight=10)`
2. `SetWeight(nonce=6, weight=7)`
3. `SetWeight(nonce=7, weight=11)`

Rather than requiring all the messages to be delivered, we can just deliver the last message and the weight will be set correctly.